### PR TITLE
ON-43061 # Added processInjectableResource to submissionService

### PR DESCRIPTION
**Pull request by mymattcarroll**

### CODE REVIEW

File: src/replaceCustomValues.ts [hunk 1]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+ * @param injector A function to inject values, this allows custom formatters to
+ *   be used. Return `undefined` to prevent the injection from recursively
+ *   continuing.
+ * @param replacer An optional function to replace nested injectables when
+ *   creating multiple resources from repeatable sets. Only required if the
+ *   `resource` param is not a `string`.
+ * @returns
+ */
+export function processInjectableResource<T>(
+  resource: T,
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:779]: I don't think the concept of a `resource` here is very comprehensible (in terms of all the docs). I think we need different terminology.

File: src/replaceCustomValues.ts [hunk 2]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+    resource: T,
+    submission: SubmissionTypes.S3SubmissionData['submission'],
+    formElements: FormTypes.FormElement[],
+  ) =>
+    | [injectedText: string, resourceKey: string, newResource: T]
+    | string
+    | undefined,
+  replacer: (resource: T, replaceAll: (resourceText: string) => string) => T = (
+    resource,
+    replaceAll,
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:792]: 4 args should be an object IMO.

File: src/replaceCustomValues.ts (hunk 1 above)
Comment (mymattcarroll) [reply src/replaceCustomValues.ts:779]: Im very open to proposals :)

File: src/replaceCustomValues.ts (hunk 2 above)
Comment (mymattcarroll) [reply src/replaceCustomValues.ts:792]: This is actually 5 args... 😱

File: tests/submissionService.test.ts [hunk 3]
```diff
@@ -585,3 +590,334 @@ describe('replaceInjectablesWithSubmissionValues()', () => {
...
+                label: 'Name',
+                readOnly: false,
+                required: false,
+                conditionallyShow: false,
+                requiresAllConditionallyShowPredicates: false,
+                isElementLookup: false,
+                isDataLookup: false,
+              },
+            ],
+          },
```
Comment (jdawg093) [inline tests/submissionService.test.ts:818]:
```suggestion
```
Assuming you don't want 2 of these

File: src/replaceCustomValues.ts [hunk 4]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+  }
+
+  const [text, resourceKey, newResource] =
+    typeof injectorResult === 'string'
+      ? [injectorResult, injectorResult, injectorResult as T]
+      : injectorResult
+
+  // Find nested form elements
+  const matches: Map<string, boolean> = new Map()
+  matchElementsTagRegex(
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:809]: Is the idea here that non nesteds have already been replaced, so the only remaining ones to look for are nested?

Comment (mymattcarroll) [reply src/replaceCustomValues.ts:809]: correcto.

FYI, this is existing code from apps-react.

Comment (mymattcarroll) [reply src/replaceCustomValues.ts:809]: https://github.com/oneblink/apps-react/blob/master/src/services/injectableOptions.ts#L10

File: src/replaceCustomValues.ts [hunk 5]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+ * @param replacer An optional function to replace nested injectables when
+ *   creating multiple resources from repeatable sets. Only required if the
+ *   `resource` param is not a `string`.
+ * @returns
+ */
+export function processInjectableResource<T>(
+  resource: T,
+  submission: SubmissionTypes.S3SubmissionData['submission'],
+  formElements: FormTypes.FormElement[],
+  injector: (
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:782]: Can this be called something like `replaceRootInjectables`

File: src/replaceCustomValues.ts [hunk 6]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+    resource: T,
+    submission: SubmissionTypes.S3SubmissionData['submission'],
+    formElements: FormTypes.FormElement[],
+  ) =>
+    | [injectedText: string, resourceKey: string, newResource: T]
+    | string
+    | undefined,
+  replacer: (resource: T, replaceAll: (resourceText: string) => string) => T = (
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:790]: And this be called something like `replaceNestedInjectables` or `replaceRepeatableSetChildrenInjectables`

File: src/replaceCustomValues.ts [hunk 7]
```diff
@@ -694,3 +694,178 @@ export function replaceInjectablesWithSubmissionValues(
...
+ * @param formElements The form elements to process
+ * @param injector A function to inject values, this allows custom formatters to
+ *   be used. Return `undefined` to prevent the injection from recursively
+ *   continuing.
+ * @param replacer An optional function to replace nested injectables when
+ *   creating multiple resources from repeatable sets. Only required if the
+ *   `resource` param is not a `string`.
+ * @returns
+ */
+export function processInjectableResource<T>(
```
Comment (Zaxist) [inline src/replaceCustomValues.ts:778]:
```suggestion
export function replaceInjectablesInCustomResource<T>(
```

**Pull request merged**